### PR TITLE
add --set-shader and overhaul shader loading logic

### PR DESCRIPTION
**Pull request by LazyBumHorse**
## Issue Description

The way shaders were set and persist had a lot of inherent issues and was very unintuitive.

When loading a core, core shader presets (aka auto-shaders) would be loaded, if there are none, the `video_shader` setting would be used.

The main issue of the `video_shader` setting is how easily it is set and how persistent it is

 - If "Load Shader" in the shader menu was used, it would persist via the `video_shader` setting
 - This also happens when using "Apply Shader"
 - Because it is a setting, it will persist even on restarting RetroArch
 - There is *no way* to clear `video_shader` without editing the config manually

Historically this setting has been causing many issues because incompatible shaders would be loaded, it became so bad the setting had to be cleared via a rudimentary compatibility check (which was later replaced) and it had to be blacklisted from config overrides.

It sounded useful for setting shaders via the `--append-config` CLI option but because core presets had higher priority, it failed to deliver.

### Smaller Issues being adressed

 - There were many cases where the two shader menus weren't syncing up properly
 - If an auto-shader failed loading, the shader menus would still show it
 - Shader presets loaded via the menu don't persist through video reinit if an auto-shader exists
 - The network command interface `SET_SHADER` command wouldn't set the shader if compiled without menu

## Changes

The `video_shader` setting is completely removed.
It's place takes the `--set-shader` CLI option, which acts as a global auto-shader with highest priority (unaffected by the `auto_shaders_enable` setting)

Shaders that are loaded
 - as an auto-shader
 - through the shader menus
 - via the network command interface `SET_SHADER` command

persist until a core is unloaded.

This means they now have to be saved *explicitly*.

If a shader could not be set, it will basically disable shaders (without touching the `video_shader_enable` setting).

Both loaded and "disabled" shaders always persist through video reinit.

### Further Changes

menu_shader.c would save an applied preset to `<configname>.<type>p` (e.g. retroarch.slangp) or `menu.<type>p` as a fallback.
This looks like a remnant of the pre-auto-shaders era and has thus been removed, it will now always write to `retroarch.<type>p`

menu_shader.c would load `menu.<type>p` as a fallback when an auto-shader is not supported - but without testing if those are supported.
This logic has been completely removed.

`video_shader_read_conf_preset` contained a `command_event(CMD_EVENT_SHADER_PRESET_LOADED, NULL)` which ultimately informs the UI of updates.
This is highly unintuive and has been moved to the appropriate sections.

 - Remove old shader type fall-back logic in `menu_shader_manager_apply_changes` (superseeded by #8910)
 - Optimize `video_shader_any_supported`
 - `video_shader_get_type_from_ext` can be called with `is_preset` = `NULL`

### CLI Use

Using an absolute path:
`retroarch -L <core> --set-shader "D:\RetroArch\shaders\shaders_glsl\blurs\kawase_blur_5pass.glslp" <content>`

Or relative to `video_shader_dir`:
`retroarch -L <core> --set-shader "shaders_glsl\blurs\kawase_blur_5pass.glslp" <content>`

Turning off auto-shaders:
`retroarch -L <core> --set-shader "" <content>`

Note: Won't work due to CLI syntax:
`retroarch -L <core> --set-shader <content>`

### Network Command Interface Use

Using an absolute path:
`retroarch --command "SET_SHADER D:\RetroArch\shaders\shaders_glsl\blurs\kawase_blur_5pass.glslp"`

Relative to `video_shader_dir`:
`retroarch --command "SET_SHADER shaders_glsl\blurs\kawase_blur_5pass.glslp"`

Clears current shader:
`retroarch --command SET_SHADER`

## Related Issues

Fixes https://github.com/libretro/RetroArch/issues/5398

## Unresolved Issues

The desktop shader menu can crash when closing content sometime after

> [INFO] [Qt]: Reloading shader parameters.

## Future Plans

Currently, because of the design of video driver `set_shader` and the shader menu, every loaded shader is actually read and parsed twice, once in the shader backend and a second time in the shader menu.
Similarly, when a shader is set from the menu, it'll be writen to a preset file which will then be immediately parsed again by the shader backend.

The video driver `set_shader` should really take a shader (preset) object, so we'd need to read and parse only once.
This is easier said than done however.

### CODE REVIEW

**Comment by rsn8887** [comment]
You failed to explain how I can set a global default shader preset now using just the menu? E.g. how do I get the previous functionality?

**Comment by LazyBumHorse** [comment]
You currently can't, the next best thing is using a core shader preset, I hope that's not too inconvenient.
I might add global shader presets in the future as well - implementing them is not hard, but changing all the UI to reflect that is a bit daunting.

**Comment by rsn8887** [comment]
So you are saying with your PR here I now have to manually set the shader for each core? And we’re js it saved? In the core options files? I find that quite inconvenient, but as long as it is saved and I do t have to do it by hand every time I guess it is ok.

**Comment by LazyBumHorse** [comment]
> I now have to manually set the shader for each core

Yeah, it's luckily a one time thing.

> we’re js it saved? In the core options files?

The presets are written in the shader directory under `shaders/presets/<core name>/`

**Comment by rsn8887** [comment]
The preset that is written is different from the preset I loaded.

It uses “../../shaders_glsl/“ instead of “./“. This is an invalid path because there’s no shaders_glsl folder only a shaders folder.

**Comment by Tatsuya79** [comment]
I'm not sure, is there still a shaders/retroarch.<cgp/glslp/slangp> used at all now?
No more shaders can be added in core overrides? (like config\gambatte\gambatte.cfg)

**Comment by rsn8887** [comment]
Was this PR and the Core preset functionality tested at all I wonder?

**Comment by Tatsuya79** [comment]
Yeah I feel this was merged a bit too fast! :sweat_smile: 
Time for testing I guess...

**Comment by LazyBumHorse** [comment]
@rn8887 The path is relative to the file itself, so "../../" goes up to the shader directory, "shaders_glsl" goes further in where the actual shader files are.
This PR did not change a thing about how shader presets are saved and, yes, everything is tested and functioning... please don't be so demeaning

@Tatsuya79 `shaders/retroarch.<cgp/glslp/slangp>` is used when you apply a shader from the menu, it's currently still needed because of design issues (see "Future Plans")

**Comment by rsn8887** [comment]
It works for me actually. It was just that older Retroarch packages used a shaders folder only, but newer versions use a shaders/shaders_glsl subfolder that was my confusion. I think I am kind of on with the changes here, because shaders were confusing before. E.g it was a quickmenu option but it was applied globally. This had always bothered me because everything else in quickmenu is per-core.

**Comment by LazyBumHorse** [comment]
I aggree however, that it was merged too fast... didn't expect such a backlash

**Comment by rsn8887** [comment]
Sorry it was mostly my own confusion and backlash. I think it is ok we just have to tell users that they now have to click „save core shader preset...“ explicitly and for each core separately if they want their shader choice to stick. This is a pretty drastic change from previous functionality where you set it once in any core, and then it was active in all cores.

**Comment by rsn8887** [comment]
It is more logical this way though. Although I would have preferred for the shader menu to just be moved to settings->video where it belongs IMO. I never understood why shader settings could only be accessed via quickmenu.

**Comment by Tatsuya79** [comment]
Not a backlash! just trying to understand what it's about.
We all know the issues with the shaders system, I hate it when RA sets it globally silently sometimes.

**Comment by LazyBumHorse** [comment]
Yeah, I'm sorry as well, I got quite flustered there, it's not a "backlash" but legitimate critisism.

It would probably have been better to add global presets first.

**Comment by rsn8887** [comment]
Honestly I am very glad somebody is taking on this issue it had also bothered me before. But then I got used to the fact that it was always automatically global and forgot about it :)

**Comment by LazyBumHorse** [comment]
Oops, missed this one:
> No more shaders can be added in core overrides? (like config\gambatte\gambatte.cfg)

@Tatsuya79 The answer is no, but you can save a core preset for gambatte.
Does this limit functionality in some way?

**Comment by Tatsuya79** [comment]
Apart from forcing me to convert all mine to presets I don't think so. :sweat_smile: 
I'm doing that right now to see if I bump into any limitation or bug.

**Comment by Tatsuya79** [comment]
I can't get out of the "load preset" navigation in xmb when pushing "back".
If you push F1 twice you go back to the root menu "Quick menu".

**Comment by LazyBumHorse** [comment]
Just for future reference: Caused by this https://github.com/libretro/RetroArch/commit/cb9a9e9525f7ec0e68b94172e2214a9fc6975a71, fixed by this https://github.com/libretro/RetroArch/commit/dec409bf17a3808e9b69043f1e9bc9783e944f17

**Comment by jdgleaver** [comment]
Excuse the delayed response - I only just logged on and saw this...

>>    No more shaders can be added in core overrides? (like config\gambatte\gambatte.cfg)
>
>@Tatsuya79 The answer is no, but you can save a core preset for gambatte.
>Does this limit functionality in some way?

This absolutely limits functionality. Let me explain:

I currently use core/content directory config overrides to set all my shaders. So, as one example, I have 20+ config overrides all pointing to the same heavily customised CRT shader. I also regularly tweak/refine my shader settings. Before this PR, I just had to tweak/maintain a single preset file (in my own shader repository), and all cores/content would receive the updates.

With this PR, I now have to manually create a new shader preset for each core/content directory. While this in itself is neither more or less inconvenient than creating config overrides, it suffers from a huge problem: these new presets are completely detached from the source file. Now, if I want to tweak a shader, I can't just update one file in my shader repo - I have to manually update 20+ preset copies. It might take me 30-40 attempts to adjust settings to my satisfaction - that's 30-40 times that I have to update 20+ files. On a desktop PC this is unmanageable. On a platform like Android it's impractical to the point of impossibility.

Now I appreciate that most users just blindly select whatever is provided in the 'official' shader repo, but even here there is a problem - if a preset is updated in the official repo and the user grabs it via the online updater, none of their per-core presets will get updated. After each update, users will have to re-create all their shader overrides 'just in case' something has changed in the source shader preset.

TBH, I am astonished that such an important/fundamental piece of functionality has been so casually discarded.

**Comment by LazyBumHorse** [comment]
@jdgleaver This can be solved by adjusting how presets work.
One idea is to add a `#include <preset path>` directive which points to another preset file to be loaded.
The only negative I can come up with this is a double read, so it'll hurt performance a bit, though maybe there's a better way to store all auto-presets.

It could work like this:
If a preset is loaded in the menu and then saved without having changed it, `#include` could be used.

**Comment by LazyBumHorse** [comment]
> I have 20+ config overrides all pointing to the same heavily customised CRT shader

Another thought: Wouldn't it be easier to make it a global preset and then "exclude" everything else by using a stock preset?

**Comment by LazyBumHorse** [comment]
Another obvious remedy would be to add another setting to override shaders just like --set-shader does, but it can't be called `video_shader`.
This is probably the easiest fix.

**Comment by jdgleaver** [comment]
>@jdgleaver This can be solved by adjusting how presets work.
>One idea is to add a #include <preset path> directive which points to another preset file to be loaded.
>The only negative I can come up with this is a double read, so it'll hurt performance a bit, though maybe there's a better way to store all auto-presets.
>
>It could work like this:
>If a preset is loaded in the menu and then saved without having changed it, #include could be used.

This is a truly excellent solution. It not only replaces the old config override functionality with something better, but it also solves all bugbears with the existing 'save core/content directory/game preset' implementation - i.e. all preset overrides will automatically link to the original source file unless the user specifically changes something. This would be perfection as far as I'm concerned.

The performance impact would be negligible, since an 'include-type' preset would only contain a single line, and the file would only be loaded when the user launches content. From my tests with runtime log files, this kind of resource usage is no issue at all (even on garbage-tier devices)

**Comment by LazyBumHorse** [comment]
> This would be perfection as far as I'm concerned.

I honestly did not expect such a great reaction :smile: 

I guess it's settled then, `#include` it is!

**Comment by jdgleaver** [comment]
Heh - I didn't expect such an awesome solution!  :)

What with your existing overhaul + this new `#include` functionality, you've single-handedly fixed all the shader issues that have plagued RetroArch for years.

So cool. Thank you so much for this!

**Comment by LazyBumHorse** [comment]
> you've single-handedly fixed all the shader issues that have plagued RetroArch for years

That's too much of an honor, I just tried to make the logic a bit saner.
I mean it's still not as good as sliced bread :grin:

> Thank you so much for this!

Thank you for the encouraging words!
Alright, back to the code

**Comment by LazyBumHorse** [comment]
@jdgleaver 
So I tried implementing the loading function which handles the `#import` and falls back to regular (config-)loading and got it to work almost immediately for absolute paths.
I then improved it further, added some debug code and at some point got super confused because the debug code would not be called - but the shader still loaded correctly? Even when the `#import` path is relative?

Turns out, the regular old config loading function does actually handle `#import`!
And it even works with relative paths.

In other words, you can create presets with `#include` *right now*.
I thought only the shader parser had `#include` support, what a nice surprise!

Though my code now needs to land in the garbage, poor thing.

**Comment by jdgleaver** [comment]
Oh my goodness - how could I not have realised that?!?

Amazing...

Well, I'm sorry you wasted your time, but it's a good result nonetheless  :)

So I guess the only remaining thing is to implement the *saving* of shader presets using the `#import` directive. Would you like to look at this, or shall I? (I'm happy either way - this is a cool thing, and it's piqued my interest)

**Comment by LazyBumHorse** [comment]
I can do the saving, I know where all the stuff is because of the global preset support, might get a bit messy because of the additional state, but let's see.

**Comment by jdgleaver** [comment]
You are a scholar and a gentleman.  :)

**Comment by Awakened0** [comment]
> 
> 
> Another obvious remedy would be to add another setting to override shaders just like --set-shader does, but it can't be called `video_shader`.
> This is probably the easiest fix - at least for the config override issue.

Maybe something like a read-only version of video_shader that could only be manually edited into configs? Organizationally, I like having my shader overrides included with the other overridden settings instead of being cordoned off in shaders/presets.

**Comment by LazyBumHorse** [comment]
> Maybe something like a read-only version of video_shader that could only be manually edited into configs?

A setting has at least three issues though:

 - It can only point to one shader type
 - [This issue](https://github.com/libretro/RetroArch/issues/1718) would "leak" the setting into the main config when --append-config is used
 - The semantics are _very_ different from auto-shader

It's not as versatile and potentially dangerous, let me explain:
The difference in semantics is that it is an override, which means, for example, if you already have a bunch of game presets and decide to use a core override, then the game presets will never be loaded.
Even worse, if the setting leaks in the main config, no shader preset will work anymore.

The semantics clash, it's basically an either-or and from a functionality standpoint, presets are superior.
This is basically why this was my last suggestion.

> I like having my shader overrides included with the other overridden settings instead of being cordoned off in shaders/presets.

If we ignore the different semantics for now, I can see how this might be easier when manually editing overrides, but I personally I don't like the redundancy of having 2 features doing the same thing, just in different places.

In theory it's possible to add overrides for concrete auto-shaders as settings, e.g. `"auto_shader_Snes9x_Snes9x"` for a Snes9x core preset override (basically just mirroring the folder/file structure auto-shaders use) but it's... not exactly elegant, to keep it tame.
As a single setting e.g. something like `"auto_shaders" = "Snes9x_Snes9x,one_preset.slangp;global,another_preset.slangp"` (using `,` and `;` as separators here) it would be a bit nicer, but it also wouldn't work with config overrides at all, because you can only override the whole setting.

**Comment by Awakened0** [comment]
I never had leaking issues since I use config_save_on_exit = "false" in my primary config. I always edit the main config or my overrides to change shaders. I really only use the menu for loading shaders I'm checking out for the first time or fiddling with parameters (which I'll edit the default for directly in the shader passes if I want to change them).

But yeah, I was also thinking adding that might be too much, since how many ways do you really need to set shaders.
